Prepare console build as early as possible

andytson-inviqa opened the pull request:
allows developers to start seeing the full codebase on the host as soon as it's available after build and introduces a new command just to get to that state `ws enable console`

andytson-inviqa committed:
Sync and resume mutagen as fast as possible

and avoid wiping it once done unless ws destroy'ed

allows developers to start seeing the full codebase on the host as soon as it's available, and avoids unnecessary repeated deletion and recreation of the mutagen sync volume on errors at first startup

andytson-inviqa commented on src/_base/harness/scripts/enable.sh.twig:
```diff
@@ -15,34 +15,40 @@ main()
-        passthru ws mutagen start
+    if [ "$USE_MUTAGEN" = yes ]; then
+        if [ "$IS_SYNCED" = no ]; then
+            passthru ws mutagen start
```
could do with renaming start as it's destructive, so `recreate` or making it an option `... start --recreate`

andytson-inviqa commented:
> allows developers to start seeing the full codebase on the host as soon as it's available after build

aka composer has finished install, do I need to wait for the rest of the init/startup process to open the codebase in an IDE? (a bigger problem also on `ws create` for spryker, where the demoshop is only in the container)

> unnecessary repeated deletion and recreation of the mutagen sync volume on errors at first startup

aka say jenkins fails to start, or mysql isn't available, why do I need to wait 5 minutes again to recreate and flush the project volume because of that?

andytson-inviqa commented:
clicked the wrong button :)

kierenevans commented:
Do we even need to pause and resume mutagen at all during setup anymore?
When we removed docker-sync in #530, we just re-worded the comment to not mention docker-sync as we didn't want to change too much.

There is definitely some CPU overhead to having mutagen syncing file changes to the host as they happen inside the container, but perhaps more modern macbooks are better at it.

If removing the pause/resume, need to check that the sync container actually exists or we will reintroduce a similar problem to #67.

andytson-inviqa committed:
Further reduce time to considering app build synced

By starting the console and building as soon as image is built

andytson-inviqa committed:
Start dynamic services up at the same time, matching static

It would look like the only reason they were delayed was because of the previous app build being after them

andytson-inviqa committed:
Match grep change in remaining place

andytson-inviqa commented:
>Do we even need to pause and resume mutagen at all during setup anymore?

I'm not sure, but also not confident to try it out just on the project I'm on when it might more affect magento2.

> perhaps more modern macbooks are better at it.

My laptop's the same as it was back then :)


Latest change I've done has changed this even further, as I realised while I was waiting for a Go dependency and some other things to build sequentially before I was able to consider the volume safe from wiping. On the plus side the workflow for the dynamic and static builds look much similar now, just that the `app build` is done on started container vs in the image build process

andytson-inviqa committed:
Further refactor common steps between static and dynamic builds

andytson-inviqa committed:
Apply the concept to the console image itself

andytson-inviqa committed:
Expose direct enabling of console

For potential use for developers and later jenkins usage

andytson-inviqa commented:
going to rethink this a little now we've got a PR targeting the sync volume issue itself. I'd prefer not to introduce another flag file but still can have console started earlier. Just perhaps need ws enable to allow more graceful re-runs before fully built than to throw everything away

andytson-inviqa committed:
Avoid stopping console with ws enable if considered built

Avoids the loss of container data between `ws enable console` and a later `ws enable`

andytson-inviqa committed:
Remove console-build flag on rebuild

kierenevans commented on src/_base/harness/scripts/enable.sh.twig:
```diff
@@ -1,56 +1,98 @@
+        if [ "$HAS_FLAG" = no ]; then
+            # remove all services, keeping console if considered built
+            if [ -f .my127ws/.flag-console-built ]; then
+                if [ -n "$(grep -v -Fxe console <<< "$SERVICES_CREATED")" ]
```
```suggestion
                if [ -n "$(grep -v -Fxe console <<< "$SERVICES_CREATED")" ]; then
```

kierenevans committed:
Fix syntax issue

andytson-inviqa commented:
hmm, something's still stopping console still

edit: ah I see only dynamic was using the flag. static doesn't need it as the container created is enough to know it's built)

andytson-inviqa committed:
Fix issues with checking which services are running

and avoid using the flag for static builds
